clos.md

```diff
@@ -2,1655 +2,1556 @@
-    which takes keyword arguments `:x` `:y` and
-    `:z` (all defaulting to `nil` if not
-    supplied). Every time you call this function a new `point`
-    is allocated and returned.
+This gives us a CLOS type (or class) called `person` and two slots,
+named `name` and `lisper`.
 
-*   Any object returned by `make-point` will be of type
-    `point`, and will respond enthusiastically to the
-    <cite>predicate</cite> `point-p`.
+~~~lisp
+(class-of p1)
+#<STANDARD-CLASS PERSON>
 
-*   Setfable <cite>accessors</cite> `point-x`,
-    `point-y` and `point-z` can be used to read and
-    modify the slots of any `point` object.
+(type-of p1)
+PERSON
+~~~
 
-*    A shallow <cite>copier</cite>,
-    `copy-point`.
+The general form of `defclass` is:
 
-Structures can have any number of slots, from zero up (to some
-    implementation-defined limit, e.g. 254 in LispWorks for Windows) and -
-    as with lists and general vectors - the slots can hold any values.
+```
+(defclass <class-name> (list of super classes)
+  ((slot-1
+     :slot-option slot-argument)
+   (slot-2, etc))
+  (:optional-class-option
+   :another-optional-class-option))
+```
 
-In this example, **note** the form in which structures
-    are printed by default, and which can be parsed by the lisp
-    reader.
+So, our `person` class doesn't explicitely inherit from another class
+(it gets the empty parentheses `()`). However it still inherits by default from
+the class `t` and from `standard-object`. See below under
+"inheritance".
 
+We could write a minimal class definition without slots options like this:
 
 ~~~lisp
-CL-USER 1 > (defstruct point
-              x
-              y
-              z)
-POINT
+(defclass point ()
+  (x y z))
+~~~
 
-CL-USER 2 > (defun distance-from-origin (point)
-              (let* ((x (point-x point))
-                     (y (point-y point))
-                     (z (point-z point)))
-                (sqrt (+ (* x x) (* y y) (* z z)))))
-DISTANCE-FROM-ORIGIN
+or even without slots specificiers: `(defclass point () ())`.
 
-CL-USER 3 > (defun reflect-in-y-axis (point)
-              (setf (point-y point)
-                    (- (point-y point))))
-REFLECT-IN-Y-AXIS
+## Creating objects (make-instance)
 
-CL-USER 4 > (setf my-point (make-point :x 3 :y 4 :z 12))
-#S(POINT X 3 Y 4 Z 12)
+We create instances of a class with `make-instance`:
 
-CL-USER 5 > (type-of my-point)
-POINT
+~~~lisp
+(defvar p1 (make-instance 'person :name "me" ))
+~~~
 
-CL-USER 6 > (distance-from-origin my-point)
-13.0
+It is generally good practice to define a constructor:
 
-CL-USER 7 > (reflect-in-y-axis my-point)
--4
+~~~lisp
+(defun make-person (name &key lisper)
+  (make-instance 'person :name name :lisper lisper))
+~~~
 
-CL-USER 8 > my-point
-#S(POINT X 3 Y -4 Z 12)
+This has the direct advantage that you can control the required
+arguments. You should now export the constructor from your package and
+not the class itself.
```

Comment:
Perhaps an example ASDF entry here would help explain how this is done, and emphasise the distinction between a constructor and the class.

Reply:
Not sure what to add, and I wouldn't add a package definition which exports symbols. I can remove the last sentence if it's more confusing than helping, as it seems to be.